tests/testthat/test-unused_import_linter.R

```diff
@@ -0,0 +1,23 @@
+test_that("unused_import_linter lints as expected", {
+  linter <- unused_import_linter()
+  expect_lint("library(dplyr)\ntibble(a = 1)", NULL, linter)
+  # SYMBOL_FUNCTION_CALL usage is detected
+  expect_lint("library(tidyverse)\ntibble(a = 1)", NULL, linter)
+  # SYMBOL usage is detected
+  expect_lint("library(dplyr)\ndo.call(tibble, args = list(a = 1))", NULL, linter)
+  # SPECIAL usage is detected
+  expect_lint("library(magrittr)\n1:3 %>% mean()", NULL, linter)
+
+  # Missing packages are ignored
+  expect_lint("library(not.a.package)\ntibble(a = 1)", NULL, linter)
+  # SYMBOL calls with character.only = TRUE are ignored, even if the argument is a package name
+  expect_lint("library(dplyr, character.only = TRUE)\n1 + 1", NULL, linter)
+
+  msg <- rex::rex("package 'dplyr' is attached but never used")
+
+  expect_lint("library(dplyr)\n1 + 1", msg, linter)
+  expect_lint("require(dplyr)\n1 + 1", msg, linter)
+  expect_lint("library('dplyr')\n1 + 1", msg, linter)
+  # ignore namespaced usages
```

Comment:
Don't think I agree with this behavior. Some places (including at google, for our own peculiar reasons) encourage ns-qualified usage everywhere, for example.

At a minimum, let's have it set by an option?

Reply:
See also #453

Reply:
Are the lines improperly referenced here?
IIUC you want a new argument `allow_ns_usage = FALSE`?

IMO if you require a package and only use it by namespace, you shouldn't attach it, i.e. use `requireNamespace("pkg")` instead of `library(pkg)`.

Reply:
> IMO if you require a package and only use it by namespace, you shouldn't attach it, i.e. use requireNamespace("pkg") instead of library(pkg).

I see your point but a lot of people apparently disagree. The philosophy is "always ns-qualify non-base usage"; `library()` calls are added at the top to make the dependencies of the script more apparent (if `ns::` is only used once and on line 458, it can be hard to tell without a thorough scouring of the code that `ns` is a requirement -- especially if `ns::` is in a branch of the code that isn't necessarily executed except under certain parameterizations/conditions). This is very similar to C++.

Moreover, `library()` might be run at the top to induce intended side effects of `.onAttach()`. We can't tell statically if removing `library(foo)` will break the script because `.onAttach()` isn't run.

So overall, if we want to throw a lint in scripts with `library(ns)` where `ns` is only used in `ns::` form, that should only be an option that's off by default, IMO.

Reply:
I think `library()` is more like `using namespace ...`, which should be used sparingly and substituted by a non-invasive `requireNamespace()`.
Otherwise, the R search path is unnecessarily polluted, which can cause name clash issues.

https://towardsdatascience.com/attach-packages-mindfully-in-r-b9c8151b3fb4

Regarding side-effects, that's why the package exceptions are provisioned. Maybe we need more by default?

I'd be happier with a default of `FALSE` for this reason, but willing to concede the default of `allow_ns_usage` if you really think it's better for most users.

Reply:
> the R search path is unnecessarily polluted, which can cause name clash issues

hence always using `::`

> really think it's better for most users

That we've gotten #453 and #459 already tells me `TRUE` is the more common case... I just put it to a poll here:

https://twitter.com/michael_chirico/status/1519187094341517312

Hopefully we can get a decent # of responses. Let me know if the poll wording sounds fair.

Reply:
I think many people don't know about `requireNamespace()`.
Maybe which is better when referring by ns::, library or requireNamespace?

Reply:
`requireNamespace()` in particular got the most pushback:

https://twitter.com/michael_chirico/status/1519197746774769666

Overall using `library()` got the most support, but not overwhelmingly so (roughly 53-47 as of now): https://twitter.com/michael_chirico/status/1519187094341517312

Reply:
Having read the comments on twitter I also see (the same) arguments against library(). The critique against `requireNamespace()` was that it doesn't stop, which I somehow though it would 😅
Seems `loadNamespace()`, which does stop for missing packages (?), would have been the correct replacement.

So I'm still convinced `FALSE` should be default.

Reply:
That sounds like a good compromise! I would still leave it as an option but can accept a `FALSE` default.

Reply:
`requireNamespace()` does stop. `requireNamespace(quitely = TRUE)` does not stop.